src/pgp.ts

```diff
@@ -0,0 +1,92 @@
+import { createReadStream, promises as fs } from "fs";
+import * as openpgp from "openpgp";
+import * as path from "path";
+import { Readable } from "stream";
+import * as vscode from "vscode";
+import { errToStr } from "./api-helper";
+
+export type Key = openpgp.Key;
+
+export enum VerificationErrorCode {
+	/* The signature does not match. */
+	Invalid = "Invalid",
+	/* Failed to read the signature or the file to verify. */
+	Read = "Read",
+}
+
+export class VerificationError extends Error {
+	constructor(
+		public readonly code: VerificationErrorCode,
+		message: string,
+	) {
+		super(message);
+	}
+
+	summary(): string {
+		switch (this.code) {
+			case VerificationErrorCode.Invalid:
+				return "Signature does not match";
+			default:
+				return "Failed to read signature";
+		}
+	}
+}
+
+/**
+ * Return the public keys bundled with the plugin.
+ */
+export async function readPublicKeys(
+	logger?: vscode.LogOutputChannel,
+): Promise<Key[]> {
+	const keyFile = path.join(__dirname, "../pgp-public.key");
+	logger?.info("Reading public key", keyFile);
+	const armoredKeys = await fs.readFile(keyFile, "utf8");
+	return openpgp.readKeys({ armoredKeys });
+}
+
+/**
+ * Given public keys, a path to a file to verify, and a path to a detached
+ * signature, verify the file's signature.  Return true if valid, otherwise
+ * return VerificationError.
+ */
+export async function verifySignature(
+	publicKeys: openpgp.Key[],
+	cliPath: string,
+	signaturePath: string,
+	logger?: vscode.LogOutputChannel,
+): Promise<true | VerificationError> {
+	try {
+		logger?.info("Reading signature", signaturePath);
+		const armoredSignature = await fs.readFile(signaturePath, "utf8");
+		const signature = await openpgp.readSignature({ armoredSignature });
+
+		logger?.info("Verifying signature of", cliPath);
+		const message = await openpgp.createMessage({
+			// openpgpjs only accepts web readable streams.
+			binary: Readable.toWeb(createReadStream(cliPath)),
+		});
+		const verificationResult = await openpgp.verify({
+			message,
+			signature,
+			verificationKeys: publicKeys,
+		});
+		for await (const _ of verificationResult.data) {
+			// The docs indicate this data must be consumed; it triggers the
+			// verification of the data.
+		}
+		try {
+			const { verified } = verificationResult.signatures[0];
+			await verified; // Throws on invalid signature.
+			logger?.info("Binary signature matches");
+		} catch (e) {
+			const error = `Unable to verify the authenticity of the binary: ${errToStr(e)}. The binary may have been tampered with.`;
+			logger?.warn(error);
+			return new VerificationError(VerificationErrorCode.Invalid, error);
+		}
+	} catch (e) {
+		const error = `Failed to read signature or binary: ${errToStr(e)}.`;
+		logger?.warn(error);
+		return new VerificationError(VerificationErrorCode.Read, error);
+	}
+	return true;
```

Comment:
I feel weird about returning errors in js. if these were `throw`n instead we could do something like...

```ts
try {
  await verifySignature(...);
  // more happy path here
} catch (error) {
  if (error instanceof VerificationError) {
    // unhappy path here
  }
}
```

which is maybe a bit noisier than what you're currently doing, but is also a bit more "when in rome", y'know?

Reply:
You are probably right, initially I did it that way but the added nesting felt like too much, although I did break things up after that so it would not be quite so bad now.

Do you think I can get away with it if I call it `VerificationResult` instead lol